docs/upgrading.md

```diff
@@ -3,14 +3,88 @@ uri: "/docs/upgrading/"
+2. **MINOR** (e.g., `1.1.0`):
+    - A minor version increment adds new functionality in a backward-compatible manner. Users should be able to upgrade without breaking their existing code or integrations.
+    - Minor releases may include enhancements, new features, and optimizations, but they will not require users to make any changes to existing functionality.
+
+3. **PATCH** (e.g., `1.0.1`):
+    - A patch version increment is for backward-compatible bug fixes. These fixes address issues without affecting functionality or introducing breaking changes.
+    - Patch releases are safe for users to apply without concern for changes to their existing workflows.
+
+By adhering to Semantic Versioning, WPGraphQL clearly signals when users should expect breaking changes, when new features are available, and when critical bugs are fixed, helping the community make informed upgrade decisions.
+
+## Categorizing Breaking Changes
+Breaking changes will be categorized based on two dimensions: **Impact Scope** and **Effort to Upgrade**.
+
+- **Impact Scope**:
+    - Low: Affects few users (niche use cases or specific extensions).
+    - Medium: Affects a moderate number of users or common use cases.
+    - High: Affects most or all users interacting with the API.
+
+- **Effort to Upgrade**:
+    - Minimal: Requires little to no action to remain compatible.
+    - Moderate: Requires small adjustments (e.g., changing function signatures).
+    - Significant: Requires refactoring large sections of code or schema.
+
+Each breaking change will be classified using these dimensions (e.g., "High Impact, Minimal Effort").
+
+## Communication of Breaking Changes
+Breaking changes will be communicated through multiple channels to ensure broad community awareness. These include:
+
+- WPGraphQL Discord
+- Headless WordPress Discord
+- Twitter (X)
+- Facebook Advanced WordPress group
+- Mailchimp Email
+
+## Changelog and Upgrade Guide
+For any version containing breaking changes, the changelog will include:
+- A summary of the breaking changes with categorization (Impact Scope + Effort to Upgrade).
+- A highlighted link to a detailed upgrade guide.
+
+The **Upgrade Guide** will provide:
+- Step-by-step instructions to adapt to the breaking change.
+- Code examples, if applicable.
+- Recommendations for common issues during upgrades.
+
+## Release Cadence
+WPGraphQL will adopt an approach of **smaller, more frequent breaking changes**. This means:
+- **Incremental Changes**: Breaking changes will be released in smaller increments, allowing users to upgrade more easily and manage the impact with minimal disruption.
+- **Clear Communication**: Each breaking change will be categorized by **Impact Scope** and **Effort to Upgrade**, enabling users to make informed decisions about when to upgrade.
+- **Limited Version Support**: WPGraphQL will support only the last 3 major versions to reduce fragmentation, ensuring that users have clear expectations regarding version support.
```

Comment:
I was thinking that with the kind of approaches I was suggesting of adding features without breaking changes and then deprecating old systems you could probably reduce the need for supported major versions to 2. Also, just a thought, LTS?

Reply:
@justlevine ya, but I've also not really been shipping breaking changes. 

I think realistically if we ship some smaller breaking changes slightly more often, we'll need to support a handful of versions, at least for some period of time. Maybe instead of number of releases it could be period of time. 

i.e. something to the tune of "we will support the latest major version and the most prior major version for a period of 6 months" or something to that tune? 🤔 

I'm not sure what the answer should be, but I do know there will be a period of time where folks will be split on older versions as they get their projects updated, etc.

Reply:
Definitely need to think through this topic a bit more. 🤔

Reply:
> I was thinking that with the kind of approaches I was suggesting of adding features without breaking changes and then deprecating old systems you could probably reduce the need for supported major versions to 2

Some features are going to be impossible to add in a non-breaking way. . .like updating graphql-php to >15 which drops support for PHP versions below 7.4. 

I think we need to do this (upgrade the underlying GraphQL PHP) sooner than later as graphql-php 15 is now pushing 2 years old. 

Many other features that might be breaking I do think can be done in more incremental fashion. 

Perhaps we prioritize this update and any communication around this update to get to a 2.0, then work on any other potential breaking changes post-2.0 in a more incremental fashion as described, where we can introduce new features that might not be active until you opt-in, and then at a later date they become the defaults.

Reply:
understood. do what you think's right. sounds like a light touch on this doc might be best and we can always add clarity later

Reply:
Ok, I've updated to state latest 2 versions instead of latest 3. I want to encourage folks to stay updated with the latest and greatest, but I know there will be _some_ periods of fragmentation as folks allocate time to get updated. 

Some breaking changes will likely be quick updates for the masses, while some may take more time, and I don't want to fully fragment the community by not offering _some_ support for other versions, but also we don't have a massive team working on this to provide support for a matrix of versions.

Reply:
@jasonbahl per my primary feedback, I think first and foremost you need to define what "support" actually means.

Reply:
@justlevine how's something to the tune of:

```
## Support Policy

WPGraphQL will support the **last 2 major versions** to reduce version fragmentation. This means that older versions beyond this window will no longer receive official support, and users will be encouraged to upgrade to the latest version.

### What Does "Support" Mean?
"Support" includes:
- **Bug Fixes**: Active work will be done to resolve bugs and issues reported on supported versions.
- **Security Updates**: General and critical security fixes will be backported to the last 2 major versions, ensuring that these versions remain secure.
- **Compatibility**: Supported versions will be tested against newer versions of WordPress, PHP, and common extensions to ensure continued functionality.
- **Documentation**: Currently our documentation on wpgraphql.com is only setup to provide documentation for the latest version of the plugin. We _may_ explore initiatives to publish documentation for multiple versions.

### Unsupported Versions:
Versions older than the last 2 major versions will no longer receive:
- Bug fixes
- Security updates
- Compatibility testing or guarantees with new versions of WordPress, PHP, or third-party extensions

While older versions may still function, the quality and speed of support will be higher for those using the latest versions.
```

Reply:
Style-wise, love it. 
- I'd  make it explicit that it's the latest `minor.patch` that's supported for each major, and not every `major.minor`.

Regarding the policy itself: 
- the `bug fixes` make me really nervous and that's only considering future bugs over the breaking changes we already know we need (e.g. GraphQL 15, lazy descriptions, custom scalars, uniform hooks), not the confirmed ones already in the tracker. 
- Similarly, compatibility testing with WordPress we can do, but I don't think it's feasible for PHP or "common extensions".

Ultimately, you know best what resources you have at your disposal, just want to make sure you're committing with eyes open.

Reply:
> Similarly, compatibility testing with WordPress we can do, but I don't think it's feasible for PHP or "common extensions".

Right now our testing matrix tests new releases against various versions of WordPress and PHP. We don't do testing against common extensions, so perhaps I should remove that. In theory I/we do support common extensions, i.e. if you open an issue or discussion in Github and you're using something like WPGraphQL and Polylang, we will typically help you with your issue. 

My intent here was to say that if you're lingering on an old version of WPGraphQL (for whatever reason) we likely won't be supporting your use case with a common extension in the same way we would support you if you were on the latest version.